### PR TITLE
Refine mmap implement for HNSW

### LINKED ISSUE #1005: HNSW throws exception while releasing in the case of mmap enabled
the HNSW destructor calls `free` to the data part even the mmap enabled, which could be released only by `munmap`

### PR DESCRIPTION
**Pull request by yah01**
/kind bug
fix #1005

### CODE REVIEW

**Comment by yah01** [comment]
Also add tests for mmap

**Comment by yah01** [comment]
After trying, to map the whole file may be the simplest way, but we keep the graph in memory by copying it, to avoid much performance downgrade

**Comment by yah01** [comment]
Also add tests for all index support mmap

File: src/io/fileIO.h [hunk 1]
```diff
@@ -17,46 +17,57 @@
 
 namespace knowhere {
 struct FileReader {
-    int fd;
-    size_t size;
-
     FileReader(const std::string& filename, bool auto_remove = false) {
-        fd = open(filename.data(), O_RDONLY);
-        if (fd < 0) {
-            std::runtime_error("Cannot open file");
```
Comment (liliu-z) [inline src/io/fileIO.h:26]: Emmm...

File: thirdparty/hnswlib/hnswlib/hnswalg.h [hunk 2]
```diff
@@ -643,7 +646,9 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
     loadIndex(const std::string& location, const knowhere::Config& config, size_t max_elements_i = 0) {
         auto cfg = static_cast<const knowhere::BaseConfig&>(config);
 
-        auto input = knowhere::FileReader(location, true);
+        auto input = knowhere::FileReader(location);
```
Comment (liliu-z) [inline thirdparty/hnswlib/hnswlib/hnswalg.h:649]: You may want to close it somewhere.

Comment (yah01) [reply thirdparty/hnswlib/hnswlib/hnswalg.h:649]: > You may want to close it somewhere.

Line 747 closes it

File: thirdparty/hnswlib/hnswlib/hnswalg.h [hunk 3]
```diff
@@ -686,21 +691,17 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
         readBinaryPOD(input, mult_);
         readBinaryPOD(input, ef_construction_);
 
-        auto pos = input.offset();
-
         if (cfg.enable_mmap.has_value() && cfg.enable_mmap.value()) {
-            cfg.enable_mmap = true;
```
Comment (liliu-z) [inline thirdparty/hnswlib/hnswlib/hnswalg.h:692]: Emm.

**Pull request merged**